[new-parser] No whitespace between rule keyword and rule name causes problems

## Parent issue
- #5678 

## Failing tests
- `org.drools.testcoverage.functional.parser.DrlParserTest#testParserSmoke`
  - `drools-test-coverage/test-suite/src/test/resources/org/drools/testcoverage/functional/parser/drl/whitespaces.drl`

## Rule code snippet
```
rule "5" when str :String() then end

rule"6" when then end rule"7" when then end
```

`LexerHelper` does not recognize `rule` in `rule"6"` as a valid end of the previous rule's RHS and so `rule"6"` is tokenized as `RHS_CHUNK`.

## Error output
```
### parse : ANTLR4_PARSER_ENABLED = true
00:26:32.836 [main] ERROR o.d.c.k.b.impl.AbstractKieProject.buildKnowledgePackages:280 - Unable to build KieBaseModel:KieBaseModelName
InvalidExpressionErrorResult: Unable to parse consequence caused by: (line 4,col 1) Parse error. Found  "\"6\"" <STRING_LITERAL>, expected one of  "," ";" "=" "@" "["
Problem stacktrace : 
  com.github.javaparser.GeneratedJavaParser.generateParseException(GeneratedJavaParser.java:13933)
  com.github.javaparser.GeneratedJavaParser.jj_consume_token(GeneratedJavaParser.java:13778)
  com.github.javaparser.GeneratedJavaParser.BlockStatement(GeneratedJavaParser.java:5807)
  com.github.javaparser.GeneratedJavaParser.Statements(GeneratedJavaParser.java:2773)
  com.github.javaparser.GeneratedJavaParser.Block(GeneratedJavaParser.java:5744)
  com.github.javaparser.GeneratedJavaParser.BlockParseStart(GeneratedJavaParser.java:8238)
  com.github.javaparser.JavaParser.parse(JavaParser.java:125)
  com.github.javaparser.JavaParser.parseBlock(JavaParser.java:317)
  org.drools.model.codegen.execmodel.generator.DrlxParseUtil.parseBlock(DrlxParseUtil.java:500)
  org.drools.model.codegen.execmodel.generator.Consequence.rewriteConsequence(Consequence.java:229)
  org.drools.model.codegen.execmodel.generator.Consequence.createCall(Consequence.java:143)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.processRule(ModelGenerator.java:242)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.processRuleDescr(ModelGenerator.java:185)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.processRules(ModelGenerator.java:164)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.generateModel(ModelGenerator.java:130)
  org.drools.model.codegen.execmodel.processors.ModelGeneratorPhase.process(ModelGeneratorPhase.java:48)
  org.drools.compiler.builder.impl.processors.IteratingPhase.process(IteratingPhase.java:53)
  org.drools.model.codegen.execmodel.processors.ModelMainCompilationPhase.process(ModelMainCompilationPhase.java:104)
  org.drools.model.codegen.execmodel.ModelBuilderImpl.doSecondBuildStep(ModelBuilderImpl.java:114)
  org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build(CompositeKnowledgeBuilderImpl.java:125)
  org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build(CompositeKnowledgeBuilderImpl.java:109)
  org.drools.compiler.kie.builder.impl.AbstractKieProject.buildKnowledgePackages(AbstractKieProject.java:274)
  org.drools.compiler.kie.builder.impl.AbstractKieProject.buildKnowledgePackages(AbstractKieProject.java:220)
  org.drools.compiler.kie.builder.impl.AbstractKieProject.verify(AbstractKieProject.java:84)
  org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildKieProject(KieBuilderImpl.java:285)
  org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildAll(KieBuilderImpl.java:251)
  org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildAll(KieBuilderImpl.java:208)
  org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromKieFileSystem(KieUtil.java:147)
  org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromKieFileSystem(KieUtil.java:134)
  org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromResources(KieUtil.java:121)
  org.drools.testcoverage.functional.parser.DrlParserTest.testParserSmoke(DrlParserTest.java:45)


00:26:32.842 [main] ERROR o.d.testcoverage.common.util.KieUtil.getKieBuilderFromKieFileSystem:165 - KieBuilder errors: [Message [id=1, kieBase=KieBaseModelName, level=ERROR, path=/var/home/jlocker/src/github.com/kiegroup/drools/drools-test-coverage/test-suite/src/test/resources/org/drools/testcoverage/functional/parser/drl/whitespaces.drl, line=35, column=0
   text=Unable to parse consequence caused by: (line 4,col 1) Parse error. Found  "\"6\"" <STRING_LITERAL>, expected one of  "," ";" "=" "@" "["
Problem stacktrace : 
  com.github.javaparser.GeneratedJavaParser.generateParseException(GeneratedJavaParser.java:13933)
  com.github.javaparser.GeneratedJavaParser.jj_consume_token(GeneratedJavaParser.java:13778)
  com.github.javaparser.GeneratedJavaParser.BlockStatement(GeneratedJavaParser.java:5807)
  com.github.javaparser.GeneratedJavaParser.Statements(GeneratedJavaParser.java:2773)
  com.github.javaparser.GeneratedJavaParser.Block(GeneratedJavaParser.java:5744)
  com.github.javaparser.GeneratedJavaParser.BlockParseStart(GeneratedJavaParser.java:8238)
  com.github.javaparser.JavaParser.parse(JavaParser.java:125)
  com.github.javaparser.JavaParser.parseBlock(JavaParser.java:317)
  org.drools.model.codegen.execmodel.generator.DrlxParseUtil.parseBlock(DrlxParseUtil.java:500)
  org.drools.model.codegen.execmodel.generator.Consequence.rewriteConsequence(Consequence.java:229)
  org.drools.model.codegen.execmodel.generator.Consequence.createCall(Consequence.java:143)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.processRule(ModelGenerator.java:242)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.processRuleDescr(ModelGenerator.java:185)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.processRules(ModelGenerator.java:164)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.generateModel(ModelGenerator.java:130)
  org.drools.model.codegen.execmodel.processors.ModelGeneratorPhase.process(ModelGeneratorPhase.java:48)
  org.drools.compiler.builder.impl.processors.IteratingPhase.process(IteratingPhase.java:53)
  org.drools.model.codegen.execmodel.processors.ModelMainCompilationPhase.process(ModelMainCompilationPhase.java:104)
  org.drools.model.codegen.execmodel.ModelBuilderImpl.doSecondBuildStep(ModelBuilderImpl.java:114)
  org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build(CompositeKnowledgeBuilderImpl.java:125)
  org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build(CompositeKnowledgeBuilderImpl.java:109)
  org.drools.compiler.kie.builder.impl.AbstractKieProject.buildKnowledgePackages(AbstractKieProject.java:274)
  org.drools.compiler.kie.builder.impl.AbstractKieProject.buildKnowledgePackages(AbstractKieProject.java:220)
  org.drools.compiler.kie.builder.impl.AbstractKieProject.verify(AbstractKieProject.java:84)
  org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildKieProject(KieBuilderImpl.java:285)
  org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildAll(KieBuilderImpl.java:251)
  org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildAll(KieBuilderImpl.java:208)
  org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromKieFileSystem(KieUtil.java:147)
  org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromKieFileSystem(KieUtil.java:134)
  org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromResources(KieUtil.java:121)
  org.drools.testcoverage.functional.parser.DrlParserTest.testParserSmoke(DrlParserTest.java:45)
]]

java.lang.AssertionError: [[Message [id=1, kieBase=KieBaseModelName, level=ERROR, path=/var/home/jlocker/src/github.com/kiegroup/drools/drools-test-coverage/test-suite/src/test/resources/org/drools/testcoverage/functional/parser/drl/whitespaces.drl, line=35, column=0
   text=Unable to parse consequence caused by: (line 4,col 1) Parse error. Found  "\"6\"" <STRING_LITERAL>, expected one of  "," ";" "=" "@" "["
Problem stacktrace : 
  com.github.javaparser.GeneratedJavaParser.generateParseException(GeneratedJavaParser.java:13933)
  com.github.javaparser.GeneratedJavaParser.jj_consume_token(GeneratedJavaParser.java:13778)
  com.github.javaparser.GeneratedJavaParser.BlockStatement(GeneratedJavaParser.java:5807)
  com.github.javaparser.GeneratedJavaParser.Statements(GeneratedJavaParser.java:2773)
  com.github.javaparser.GeneratedJavaParser.Block(GeneratedJavaParser.java:5744)
  com.github.javaparser.GeneratedJavaParser.BlockParseStart(GeneratedJavaParser.java:8238)
  com.github.javaparser.JavaParser.parse(JavaParser.java:125)
  com.github.javaparser.JavaParser.parseBlock(JavaParser.java:317)
  org.drools.model.codegen.execmodel.generator.DrlxParseUtil.parseBlock(DrlxParseUtil.java:500)
  org.drools.model.codegen.execmodel.generator.Consequence.rewriteConsequence(Consequence.java:229)
  org.drools.model.codegen.execmodel.generator.Consequence.createCall(Consequence.java:143)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.processRule(ModelGenerator.java:242)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.processRuleDescr(ModelGenerator.java:185)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.processRules(ModelGenerator.java:164)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.generateModel(ModelGenerator.java:130)
  org.drools.model.codegen.execmodel.processors.ModelGeneratorPhase.process(ModelGeneratorPhase.java:48)
  org.drools.compiler.builder.impl.processors.IteratingPhase.process(IteratingPhase.java:53)
  org.drools.model.codegen.execmodel.processors.ModelMainCompilationPhase.process(ModelMainCompilationPhase.java:104)
  org.drools.model.codegen.execmodel.ModelBuilderImpl.doSecondBuildStep(ModelBuilderImpl.java:114)
  org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build(CompositeKnowledgeBuilderImpl.java:125)
  org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build(CompositeKnowledgeBuilderImpl.java:109)
  org.drools.compiler.kie.builder.impl.AbstractKieProject.buildKnowledgePackages(AbstractKieProject.java:274)
  org.drools.compiler.kie.builder.impl.AbstractKieProject.buildKnowledgePackages(AbstractKieProject.java:220)
  org.drools.compiler.kie.builder.impl.AbstractKieProject.verify(AbstractKieProject.java:84)
  org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildKieProject(KieBuilderImpl.java:285)
  org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildAll(KieBuilderImpl.java:251)
  org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildAll(KieBuilderImpl.java:208)
  org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromKieFileSystem(KieUtil.java:147)
  org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromKieFileSystem(KieUtil.java:134)
  org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromResources(KieUtil.java:121)
  org.drools.testcoverage.functional.parser.DrlParserTest.testParserSmoke(DrlParserTest.java:45)
]]] 
Expecting empty but was: [Message [id=1, kieBase=KieBaseModelName, level=ERROR, path=/var/home/jlocker/src/github.com/kiegroup/drools/drools-test-coverage/test-suite/src/test/resources/org/drools/testcoverage/functional/parser/drl/whitespaces.drl, line=35, column=0
   text=Unable to parse consequence caused by: (line 4,col 1) Parse error. Found  "\"6\"" <STRING_LITERAL>, expected one of  "," ";" "=" "@" "["
Problem stacktrace : 
  com.github.javaparser.GeneratedJavaParser.generateParseException(GeneratedJavaParser.java:13933)
  com.github.javaparser.GeneratedJavaParser.jj_consume_token(GeneratedJavaParser.java:13778)
  com.github.javaparser.GeneratedJavaParser.BlockStatement(GeneratedJavaParser.java:5807)
  com.github.javaparser.GeneratedJavaParser.Statements(GeneratedJavaParser.java:2773)
  com.github.javaparser.GeneratedJavaParser.Block(GeneratedJavaParser.java:5744)
  com.github.javaparser.GeneratedJavaParser.BlockParseStart(GeneratedJavaParser.java:8238)
  com.github.javaparser.JavaParser.parse(JavaParser.java:125)
  com.github.javaparser.JavaParser.parseBlock(JavaParser.java:317)
  org.drools.model.codegen.execmodel.generator.DrlxParseUtil.parseBlock(DrlxParseUtil.java:500)
  org.drools.model.codegen.execmodel.generator.Consequence.rewriteConsequence(Consequence.java:229)
  org.drools.model.codegen.execmodel.generator.Consequence.createCall(Consequence.java:143)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.processRule(ModelGenerator.java:242)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.processRuleDescr(ModelGenerator.java:185)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.processRules(ModelGenerator.java:164)
  org.drools.model.codegen.execmodel.generator.ModelGenerator.generateModel(ModelGenerator.java:130)
  org.drools.model.codegen.execmodel.processors.ModelGeneratorPhase.process(ModelGeneratorPhase.java:48)
  org.drools.compiler.builder.impl.processors.IteratingPhase.process(IteratingPhase.java:53)
  org.drools.model.codegen.execmodel.processors.ModelMainCompilationPhase.process(ModelMainCompilationPhase.java:104)
  org.drools.model.codegen.execmodel.ModelBuilderImpl.doSecondBuildStep(ModelBuilderImpl.java:114)
  org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build(CompositeKnowledgeBuilderImpl.java:125)
  org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build(CompositeKnowledgeBuilderImpl.java:109)
  org.drools.compiler.kie.builder.impl.AbstractKieProject.buildKnowledgePackages(AbstractKieProject.java:274)
  org.drools.compiler.kie.builder.impl.AbstractKieProject.buildKnowledgePackages(AbstractKieProject.java:220)
  org.drools.compiler.kie.builder.impl.AbstractKieProject.verify(AbstractKieProject.java:84)
  org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildKieProject(KieBuilderImpl.java:285)
  org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildAll(KieBuilderImpl.java:251)
  org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildAll(KieBuilderImpl.java:208)
  org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromKieFileSystem(KieUtil.java:147)
  org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromKieFileSystem(KieUtil.java:134)
  org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromResources(KieUtil.java:121)
  org.drools.testcoverage.functional.parser.DrlParserTest.testParserSmoke(DrlParserTest.java:45)
]]

	at org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromKieFileSystem(KieUtil.java:169)
	at org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromKieFileSystem(KieUtil.java:134)
	at org.drools.testcoverage.common.util.KieUtil.getKieBuilderFromResources(KieUtil.java:121)
	at org.drools.testcoverage.functional.parser.DrlParserTest.testParserSmoke(DrlParserTest.java:45)
```

--- a/drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java
+++ b/drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java
@@ -4666,4 +4666,15 @@ class MiscDRLParserTest {
         assertThat(children).isNotEmpty(); // Make sure that every child type is represented.
         assertThat(children).allSatisfy(baseDescr -> assertThat(baseDescr.getNamespace()).isEqualTo(namespace));
     }
+
+    @Test
+    public void noWhitespaceBetweenRuleKeywordAndName() {
+        final String text = "rule X when then end rule\"Y\" when then end rule'Z'when then end";
+
+        PackageDescr pkg = parseAndGetPackageDescr(text);
+
+        assertThat(pkg.getRules())
+                .map(RuleDescr::getName)
+                .containsExactly("X", "Y", "Z");
+    }
 }

--- a/drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/antlr4/LexerHelper.java
+++ b/drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/antlr4/LexerHelper.java
@@ -18,7 +18,7 @@
  */
 package org.drools.drl.parser.antlr4;
 
-import java.util.List;
+import java.util.Set;
 
 import org.antlr.v4.runtime.CharStream;
 import org.antlr.v4.runtime.IntStream;
@@ -29,9 +29,10 @@ import org.drools.drl.parser.lang.DroolsSoftKeywords;
  */
 public class LexerHelper {
 
-    private static final List<Character> semiAndWS = List.of(';', ' ', '\t', '\n', '\r');
-    private static final List<String> statementKeywordsList = List.of(ParserHelper.statementKeywords);
-    private static final List<String> attributeKeywordsList = List.of(DroolsSoftKeywords.SALIENCE,
+    private static final Set<Character> semiAndWS = Set.of(';', ' ', '\t', '\n', '\r');
+    private static final Set<Character> delimiters = Set.of(';', ' ', '\t', '\n', '\r', '"', '\'');
+    private static final Set<String> statementKeywordsList = Set.of(ParserHelper.statementKeywords);
+    private static final Set<String> attributeKeywordsList = Set.of(DroolsSoftKeywords.SALIENCE,
                                                                       DroolsSoftKeywords.ENABLED,
                                                                       DroolsSoftKeywords.NO + "-" + DroolsSoftKeywords.LOOP,
                                                                       DroolsSoftKeywords.AUTO + "-" + DroolsSoftKeywords.FOCUS,
@@ -82,7 +83,7 @@ public class LexerHelper {
         StringBuilder sb = new StringBuilder();
         while (true) {
             int la = input.LA(lookAheadCounter);
-            if (semiAndWS.contains((char) la) || la == IntStream.EOF) {
+            if (delimiters.contains((char) la) || la == IntStream.EOF) {
                 break;
             }
             sb.append((char) la);

--- a/drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/util/ParserStringUtils.java
+++ b/drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/util/ParserStringUtils.java
@@ -33,7 +33,10 @@ public class ParserStringUtils {
     public static String safeStripStringDelimiters(String value) {
         if (value != null) {
             value = value.trim();
-            if (value.length() >= 2 && value.startsWith("\"") && value.endsWith("\"")) {
+            if (value.length() >= 2 && (
+                    value.startsWith("\"") && value.endsWith("\"")
+                    || value.startsWith("'") && value.endsWith("'")
+            )) {
                 value = value.substring(1, value.length() - 1);
             }
         }

--- a/drools-test-coverage/test-suite/src/test/java/org/drools/testcoverage/functional/parser/DrlParserTest.java
+++ b/drools-test-coverage/test-suite/src/test/java/org/drools/testcoverage/functional/parser/DrlParserTest.java
@@ -34,7 +34,7 @@ public class DrlParserTest extends ParserTest {
         super(file, kieBaseTestConfiguration);
     }
 
-    @Parameters
+    @Parameters(name = "{index}: {0}, {1}")
     public static Collection<Object[]> getParameters() {
         return getTestParamsFromFiles(getFiles("drl"));
     }

